src/cli.rs

```diff
@@ -0,0 +1,764 @@
+        } else {
+            path.to_path_buf()
+        }
+    }
+
+    pub fn no_search_paths(&self) -> bool {
+        self.path.is_empty() && self.search_path.is_empty()
+    }
+
+    #[inline]
+    pub fn rg_alias_ignore(&self) -> bool {
+        self.rg_alias_hidden_ignore > 0
+    }
+
+    pub fn max_depth(&self) -> Option<usize> {
+        self.max_depth.or(self.exact_depth)
+    }
+
+    pub fn min_depth(&self) -> Option<usize> {
+        self.min_depth.or(self.exact_depth)
+    }
+
+    pub fn threads(&self) -> usize {
+        std::cmp::max(self.threads.unwrap_or_else(num_cpus::get), 1)
+    }
+
+    pub fn max_results(&self) -> Option<usize> {
+        self.max_results
+            .filter(|&m| m > 0)
+            .or_else(|| self.max_one_result.then(|| 1))
+    }
+
+    #[cfg(feature = "completions")]
+    pub fn gen_completions(&self) -> anyhow::Result<Option<Shell>> {
+        self.gen_completions
+            .map(|maybe_shell| match maybe_shell {
+                Some(sh) => Ok(sh),
+                None => guess_shell(),
+            })
+            .transpose()
+    }
+}
+
+// TODO: windows?
+#[cfg(feature = "completions")]
+fn guess_shell() -> anyhow::Result<Shell> {
+    let env_shell = std::env::var_os("SHELL").map(PathBuf::from);
+    let shell = env_shell
+        .as_ref()
+        .and_then(|s| s.file_name())
+        .and_then(|s| s.to_str())
+        .ok_or_else(|| anyhow!("Unable to get shell from environment"))?;
+    shell
+        .parse::<Shell>()
+        .map_err(|_| anyhow!("Unknown shell {}", shell))
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, ValueEnum)]
+pub enum FileType {
+    #[clap(alias = "f")]
+    File,
+    #[clap(alias = "d")]
+    Directory,
+    #[clap(alias = "l")]
+    Symlink,
+    #[clap(alias = "x")]
+    Executable,
+    #[clap(alias = "e")]
+    Empty,
+    #[clap(alias = "s")]
+    Socket,
+    #[clap(alias = "p")]
+    Pipe,
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug, ValueEnum)]
+pub enum ColorWhen {
+    /// show colors if the output goes to an interactive console (default)
+    Auto,
+    /// always use colorized output
+    Always,
+    /// do not use colorized output
+    Never,
+}
+
+impl ColorWhen {
+    pub fn as_str(&self) -> &'static str {
+        use ColorWhen::*;
+        match *self {
+            Auto => "auto",
+            Always => "always",
+            Never => "never",
+        }
+    }
+}
+
+// there isn't a derive api for getting grouped values yet,
+// so we have to use hand-rolled parsing for exec and exec-batch
+pub struct Exec {
+    pub command: Option<CommandSet>,
+}
+
+impl clap::FromArgMatches for Exec {
+    fn from_arg_matches(matches: &ArgMatches) -> clap::Result<Self> {
+        let command = matches
+            .grouped_values_of("exec")
+            .map(CommandSet::new)
+            .or_else(|| {
+                matches
+                    .grouped_values_of("exec-batch")
+                    .map(CommandSet::new_batch)
```

Comment:
side note: Oh, didn't know you use this feature.  No one has really been stepping up to advocate for it and [finish it up](https://github.com/clap-rs/clap/issues/2924)

Reply:
Is there anything I could do to help finish it?

Reply:
The linked tracking issue mentions the remaining work.